geode-core/src/main/java/org/apache/geode/cache/client/internal/pooling/ConnectionManagerImpl.java

```diff
@@ -277,148 +317,82 @@ public Connection borrowConnection(long acquireTimeout)
+      return connection;
+    }
 
-      // We need to create a connection. Reserve space for it.
-      connectionCount++;
-    } finally {
-      lock.unlock();
+    if (onlyUseExistingCnx) {
+      throw new AllConnectionsInUseException();
     }
 
-    PooledConnection connection = null;
     try {
-      Connection plainConnection = connectionFactory.createClientToServerConnection(server, false);
-      connection = addConnection(plainConnection);
+      connection = forceCreateConnection(server, Collections.emptySet());
+      if (null != connection) {
+        return connection;
+      }
     } catch (GemFireSecurityException e) {
       throw new ServerOperationException(e);
-    } finally {
-      // if we failed, release the space we reserved for our connection
-      if (connection == null) {
-        lock.lock();
-        try {
-          --connectionCount;
-          if (connectionCount < minConnections) {
-            startBackgroundPrefill();
-          }
-        } finally {
-          lock.unlock();
-        }
-      }
-    }
-    if (connection == null) {
-      throw new ServerConnectivityException(
-          "Could not create a new connection to server " + server);
     }
-    return connection;
+
+    throw new ServerConnectivityException(
+        "Could not create a new connection to server " + server);
   }
 
   @Override
-  public Connection exchangeConnection(Connection oldConnection,
-      Set/* <ServerLocation> */ excludedServers, long acquireTimeout)
+  public Connection exchangeConnection(final Connection oldConnection,
+      final Set/* <ServerLocation> */ excludedServers, final long acquireTimeout)
       throws AllConnectionsInUseException {
-    assert oldConnection instanceof PooledConnection;
-    PooledConnection newConnection = null;
-    PooledConnection oldPC = (PooledConnection) oldConnection;
 
-    boolean needToUndoEstimate = false;
-    lock.lock();
     try {
-      if (shuttingDown) {
-        throw new PoolCancelledException();
-      }
-      for (Iterator<PooledConnection> itr = availableConnections.iterator(); itr.hasNext();) {
-        PooledConnection nextConnection = itr.next();
-        if (!excludedServers.contains(nextConnection.getServer())) {
-          itr.remove();
-          try {
-            nextConnection.activate();
-            newConnection = nextConnection;
-            if (allConnectionsMap.removeConnection(oldPC)) {
-              --connectionCount;
-              if (connectionCount < minConnections) {
-                startBackgroundPrefill();
-              }
-            }
-            break;
-          } catch (ConnectionDestroyedException ex) {
-            // someone else already destroyed this connection so ignore it
-            // but remove it from availableConnections
-          }
-        }
+      PooledConnection connection =
+          findConnection(
+              (c) -> !(excludedServers.contains(c.getServer()) || oldConnection.equals(c)));
```

Comment:
I don't think the "|| oldConnection.equals(c)" is needed. The caller has already borrowed oldConnection so it will not be found by findConnection. The previous code did not do this check so does a reason exist why it is needed now?
It will not break anything but it does make the code more complicated, slows it down a bit, and could make someone think oldConnection might be in the deque

Reply:
Oh, good catch. Originally we had the old connection being put back at the end of the queue at the top of this method so we could find it again when scanning. Now that we add it after the scanning this check is not necessary.